File: src/main/java/de/tlmrgvf/drtd/gui/dialog/WaterfallDialog.java
/*
 *
 * BSD 2-Clause License
 *
 * Copyright (c) 2020, Till Mayer
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package de.tlmrgvf.drtd.gui.dialog;

import de.tlmrgvf.drtd.Drtd;
import de.tlmrgvf.drtd.dsp.window.Window;
import de.tlmrgvf.drtd.dsp.window.*;
import de.tlmrgvf.drtd.gui.component.Waterfall;
import de.tlmrgvf.drtd.gui.component.WaterfallPalette;
import de.tlmrgvf.drtd.utils.Utils;

import javax.swing.*;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;
import java.awt.*;
import java.awt.event.ItemEvent;
import java.awt.event.ItemListener;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;

public final class WaterfallDialog extends JFrame {
    private final static int MIN_BINS_EXP = 10;
    private final static int MAX_BINS_EXP = 16;
    public final static int MAX_ZOOM = 10;

    private static WaterfallDialog instance;

    private final Waterfall waterfall;
    private final JComboBox<Window> windowSelector;
    private final JSlider zoom;
    private final JSlider speedMultiplier;
    private final JSpinner fStart;
    private final JComboBox<Integer> bins;
    private final JComboBox<WaterfallPalette> palette;
    private final JCheckBox powerSpectrum;
    private final SpinnerNumberModel fStartModel;

    private boolean ignoreUpdates = false;

    private WaterfallDialog(Waterfall waterfall) {
        this.waterfall = waterfall;
        instance = this;

        setTitle("Waterfall settings");
        setResizable(false);
        setIconImages(Drtd.ICONS);
        setDefaultCloseOperation(JDialog.HIDE_ON_CLOSE);

        GridLayout gridLayout = new GridLayout(0, 2);
        JPanel rootPanel = new JPanel();
        JPanel settingsPanel = new JPanel(gridLayout);
        rootPanel.setLayout(new BoxLayout(rootPanel, BoxLayout.Y_AXIS));
        settingsPanel.setBorder(Utils.createLabeledBorder("Settings"));
        rootPanel.add(settingsPanel);
        gridLayout.setVgap(10);
        gridLayout.setHgap(10);
        rootPanel.setBorder(BorderFactory.createEmptyBorder(10, 10, 10, 10));

        windowSelector = new JComboBox<>(new Window[]{
                new RectangularWindow(),
                new BlackmanWindow(),
                new HammingWindow(),
                new HannWindow()
        });

        speedMultiplier = new JSlider(1, 10);
        speedMultiplier.setMajorTickSpacing(1);
        speedMultiplier.setPaintTicks(true);

        zoom = new JSlider(-MAX_ZOOM * 10 + 10, MAX_ZOOM * 10 - 10, 0);
        zoom.setPaintTicks(true);
        zoom.setSnapToTicks(true);
        zoom.setMajorTickSpacing(1);

        fStart = new JSpinner();
        fStartModel = new SpinnerNumberModel();
        fStart.setModel(fStartModel);
        fStartModel.setMinimum(0);
        fStartModel.setMaximum(22050);

        Integer[] powersOfTwo = new Integer[MAX_BINS_EXP - MIN_BINS_EXP + 1];
        for (int i = MIN_BINS_EXP; i <= MAX_BINS_EXP; ++i)
            powersOfTwo[i - MIN_BINS_EXP] = (int) Math.pow(2, i);

        bins = new JComboBox<>(new DefaultComboBoxModel<>(powersOfTwo));
        powerSpectrum = new JCheckBox();
        palette = new JComboBox<>(WaterfallPalette.values());

        add(rootPanel);
        settingsPanel.add(new JLabel("Window:"));
        settingsPanel.add(windowSelector);
        settingsPanel.add(new JLabel("Zoom:"));
        settingsPanel.add(zoom);
        settingsPanel.add(new JLabel("f start:"));
        settingsPanel.add(fStart);
        settingsPanel.add(new JLabel("Speed multiplier:"));
        settingsPanel.add(speedMultiplier);
        settingsPanel.add(new JLabel("Bins:"));
        settingsPanel.add(bins);
        settingsPanel.add(new JLabel("Power spectrum:"));
        settingsPanel.add(powerSpectrum);
        settingsPanel.add(new JLabel("Palette:"));
        settingsPanel.add(palette);

        updateUiFromWaterfall();

        ListenerImpl listener = new ListenerImpl();
        windowSelector.addItemListener(listener);
        zoom.addChangeListener(listener);
        fStart.addChangeListener(listener);
        bins.addItemListener(listener);
        speedMultiplier.addChangeListener(listener);
        powerSpectrum.addChangeListener(listener);
        palette.addItemListener(listener);
        zoom.addMouseListener(listener);

        pack();
        setLocationRelativeTo(Drtd.getMainGui());
        Utils.addEscapeHandler(rootPanel, this);
    }

    public static WaterfallDialog getInstance(Waterfall waterfall) {
        if (instance != null)
            return instance;

        return new WaterfallDialog(waterfall);
    }

    public void updateUiFromWaterfall() {
        ignoreUpdates = true;

        windowSelector.setSelectedItem(waterfall.getWindow());
        final var zoomFactor = (waterfall.getZoom() - 1) * 10;
        zoom.setValue((int) ((waterfall.isZoomOut() ? -zoomFactor : zoomFactor)));
        fStart.setValue(waterfall.getFrequencyOffset());
        bins.setSelectedItem(waterfall.getBins());
        speedMultiplier.setValue(waterfall.getSpeedMultiplier());
        fStartModel.setMaximum(waterfall.getSampleRate() / 2);
        powerSpectrum.setSelected(waterfall.isPowerSpectrum());
        palette.setSelectedItem(waterfall.getPalette());

        ignoreUpdates = false;
    }

    @Override
    public void setVisible(boolean b) {
        super.setVisible(b);
        setState(Frame.NORMAL);
    }

    public void updateWaterfallFromUi() {
        assert bins.getSelectedItem() != null;

        waterfall.setBins((Integer) bins.getSelectedItem());
        waterfall.setZoom(Math.abs(zoom.getValue()) / 10F + 1);
        waterfall.setZoomOut(zoom.getValue() < 0);
        waterfall.setWindow(windowSelector.getItemAt(windowSelector.getSelectedIndex()));
        waterfall.setFrequencyOffset((Integer) fStart.getValue());
        waterfall.setSpeedMultiplier(speedMultiplier.getValue());
        waterfall.setPowerSpectrum(powerSpectrum.isSelected());
        waterfall.setPalette((WaterfallPalette) palette.getSelectedItem());
    }

    private class ListenerImpl extends MouseAdapter implements ItemListener, ChangeListener {
        @Override
        public void itemStateChanged(ItemEvent itemEvent) {
            updateWaterfallFromUi();
        }

        @Override
        public void stateChanged(ChangeEvent changeEvent) {
            if (!ignoreUpdates)
                updateWaterfallFromUi();
        }

        @Override
        public void mouseClicked(MouseEvent e) {
            if (e.getClickCount() == 2)
                zoom.setValue(0);
        }
    }
}

File: src/ui/WaterfallDialog.cpp
/*
BSD 2-Clause License

Copyright (c) 2020, Till Mayer
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

1. Redistributions of source code must retain the above copyright notice, this
   list of conditions and the following disclaimer.

2. Redistributions in binary form must reproduce the above copyright notice,
   this list of conditions and the following disclaimer in the documentation
   and/or other materials provided with the distribution.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
#include "WaterfallDialog.hpp"
#include <Drtd.hpp>
#include <sstream>
#include <ui/MainGui.hpp>
#include <util/Singleton.hpp>
#include <util/Util.hpp>

using namespace Ui;

const std::array s_bins { "1024", "2048", "4096", "8192", "16384", "32768", "65536" };

WaterfallDialog::WaterfallDialog()
    : Fl_Window(100, 100, 360, 204, "Waterfall settings")
    , m_window(85, 4, 0, 25, "Window:")
    , m_zoom(m_window.x(), m_window.y() + m_window.h() + 4, w() - m_window.x() - 24, 25, "Zoom:")
    , m_reset_zoom(m_zoom.x() + m_zoom.w() + 4, m_zoom.y(), 15, m_zoom.h(), "R")
    , m_bin_offset(m_window.x(), m_zoom.y() + m_zoom.h() + 4, 150, 25, "Offset (Hz):")
    , m_speed_multiplier(m_window.x(), m_bin_offset.y() + m_bin_offset.h() + 4, w() - m_window.x() - 4, 25, "Speed:")
    , m_bins(m_window.x(), m_speed_multiplier.y() + m_speed_multiplier.h() + 4, 150, 25, "Bins:")
    , m_palette(m_window.x(), m_bins.y() + m_bins.h() + 4, 0, 25, "Palette:")
    , m_power_spectrum(m_window.x(), m_palette.y() + m_palette.h() + 4, 180, 25, "Show power spectrum") {
    double max_width = 0;

    for (auto& window : Dsp::Window::s_windows) {
        m_window.add(window.name().c_str());
        max_width = std::max(fl_width(window.name().c_str()), max_width);
    }

    m_window.size(static_cast<int>(max_width + 40), m_window.h());
    m_window.align(FL_ALIGN_LEFT);
    m_window.callback(save_to_waterfall);

    m_zoom.type(FL_HOR_NICE_SLIDER);
    m_zoom.selection_color(Util::s_amber_color);
    m_zoom.align(FL_ALIGN_LEFT);
    m_zoom.bounds(s_min_zoom, s_max_zoom);
    m_zoom.callback(save_to_waterfall);

    m_reset_zoom.callback([](Fl_Widget*, void*) {
        s_waterfall_dialog->m_zoom.value(0);
        save_to_waterfall(nullptr, nullptr);
    });

    m_bin_offset.align(FL_ALIGN_LEFT);
    m_bin_offset.callback(save_to_waterfall);

    m_speed_multiplier.type(FL_HOR_NICE_SLIDER);
    m_speed_multiplier.selection_color(Util::s_amber_color);
    m_speed_multiplier.align(FL_ALIGN_LEFT);
    m_speed_multiplier.callback(save_to_waterfall);
    m_speed_multiplier.bounds(0, 10);

    m_bins.align(FL_ALIGN_LEFT);
    for (auto bins : s_bins)
        m_bins.add(bins);
    m_bins.callback(save_to_waterfall);

    max_width = 0;
    for (auto& palette : Ui::Palette::palettes()) {
        m_palette.add(palette.name);
        max_width = std::max(fl_width(palette.name), max_width);
    }

    m_palette.size(static_cast<int>(max_width + 40), m_palette.h());
    m_palette.callback(save_to_waterfall);

    m_power_spectrum.callback(save_to_waterfall);
}

void WaterfallDialog::show_dialog() {
    auto& main_gui = Drtd::main_gui();
    s_waterfall_dialog->position(
        main_gui.x() + Util::center(main_gui.w(), s_waterfall_dialog->w()),
        main_gui.y() + Util::center(main_gui.h(), s_waterfall_dialog->h()));

    s_waterfall_dialog->load_from_waterfall(Drtd::main_gui().waterfall().settings());
    s_waterfall_dialog->set_non_modal();
    s_waterfall_dialog->show();
}

void WaterfallDialog::save_to_waterfall(Fl_Widget*, void*) {
    Waterfall::Settings new_settings;
    auto& waterfall = Drtd::main_gui().waterfall();

    new_settings.zoom = static_cast<float>(s_waterfall_dialog->m_zoom.value());
    new_settings.speed_multiplier = static_cast<u8>(s_waterfall_dialog->m_speed_multiplier.value());
    new_settings.bins = std::atoi(s_bins[static_cast<u32>(s_waterfall_dialog->m_bins.value())]);
    new_settings.power_spectrum = static_cast<bool>(s_waterfall_dialog->m_power_spectrum.value());
    s_waterfall_dialog->update_offset_spinner_limits(new_settings);
    new_settings.bin_offset = static_cast<u32>(Waterfall::translate_hz_to_x(new_settings, static_cast<Hertz>(s_waterfall_dialog->m_bin_offset.value()), waterfall.sample_rate()));

    waterfall.set_window_index(static_cast<u8>(s_waterfall_dialog->m_window.value()));
    waterfall.set_palette_index(static_cast<u8>(s_waterfall_dialog->m_palette.value()));
    waterfall.update_settings_later(new_settings);
}

void WaterfallDialog::update_offset_spinner_limits(const Waterfall::Settings& settings) {
    auto& waterfall = Drtd::main_gui().waterfall();
    m_bin_offset.step(Waterfall::hz_per_bin(settings, waterfall.sample_rate()));
    m_bin_offset.range(0, waterfall.sample_rate() / 2);
}

void WaterfallDialog::load_from_waterfall(const Waterfall::Settings& settings) {
    if (!s_waterfall_dialog.has_instance())
        return;

    auto& waterfall = Drtd::main_gui().waterfall();

    s_waterfall_dialog->m_window.value(waterfall.window_index());
    s_waterfall_dialog->m_zoom.value(settings.zoom);
    s_waterfall_dialog->m_speed_multiplier.value(settings.speed_multiplier);
    s_waterfall_dialog->update_offset_spinner_limits(settings);
    s_waterfall_dialog->m_bin_offset.value(Waterfall::translate_x_to_hz(settings, settings.bin_offset, waterfall.sample_rate()));
    s_waterfall_dialog->m_bins.value(s_waterfall_dialog->m_bins.find_item(std::to_string(settings.bins).c_str()));
    s_waterfall_dialog->m_palette.value(waterfall.palette_index());
    s_waterfall_dialog->m_power_spectrum.value(settings.power_spectrum);
}
